### ISSUE TITLE
PVC protection cannot handle terminated pods that are not deleted

### ISSUE THREAD

**Issue by pohly** [kubernetes/kubernetes#113828]
### What would you like to be added?

This came out of the discussion around https://github.com/kubernetes/kubernetes/pull/111023 which uses a concept similar to generic ephemeral volumes: since 1.22, pods are guaranteed to reach a terminal phase (failed or succeeded):https://kubernetes.slack.com/archives/C0BP8PW9G/p1668020443712729?thread_ts=1668010870.986709&cid=C0BP8PW9G

@bobbypage 
> But it’s sufficient to just look at phase after 1.22+

This means it might be possible to simplify this check: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/pvcprotection/pvc_protection_controller.go#L287

### Why is this needed?

More consistent code, dynamic resource allocation also uses podutil.IsPodTerminal ("k8s.io/kubernetes/pkg/api/v1/pod").

**Comment by msau42** [comment]
What is the proposed change?

Pods that are terminal may not have necessary completed volume teardown. So the PVC protection controller is waiting on signals that can indicate volume teardown has completed.

**Comment by alculquicondor** [comment]
/sig storage

My suggestion is two-staged:
- replace `podIsShutDown` with `podutil.IsPodTerminal`
- enhance `podutil.IsPodTerminal` to include `pod.DeletionTimestamp != nil && pod.spec.nodeName == ""`, to account for the cases where the pod was never started.

**Comment by alculquicondor** [comment]
> Pods that are terminal may not have necessarily completed volume teardown. 

I see, this implies that the pod actually has to be deleted.

However, this means that the check doesn't work for Jobs or plain Pods where there is no controller that would trigger the deletion.

**Comment by msau42** [comment]
Yeah I think to solve those use cases we need to think about other signals (or adding a new signal) we could use to indicate the volume has torn down.

@kubernetes/sig-storage-bugs

**Comment by msau42** [comment]
/retitle PVC protection cannot handle terminated pods that are not deleted

**Comment by pohly** [comment]
Removing the PVC protection is one aspect of the problem with long-lived pods that use volumes. When a generic ephemeral volume is involved, the garbage collector also won't kick in, which means the PersistentVolumeClaim doesn't get deleted and the volume remains allocated as long as the pod exists.

One could argue that this is a feature (allows debugging what the pod did with storage), but in practice it would be better to delete the volume sooner.

**Comment by msau42** [comment]
A possible (and better) signal that the PVC protection controller could use to determine volume cleanup success on a node is `Node.Status.VolumesInUse`. Volumes are removed from this list after NodeUnstageVolume has completed. This is in fact, better than Pod deletion, which only indicates that NodeUnpublishVolume has completed.

However, we need to think through possible race conditions (another Pod using the same volume gets added) if we switch to this.

**Comment by msau42** [comment]
Switching to using the Node object could also be a scalability improvement over today's method of listing all Pods. #109282

**Comment by xing-yang** [comment]
`Node.Status.VolumesInUse` is only used for NodeStageVolume but not every CSI driver supports NodeStageVolume so this list does not include all volumes. So this is another reason why we can't use this approach.
